src/java/org/apache/cassandra/service/ReplicaFilteringProtection.java

```diff
@@ -239,7 +253,25 @@ public Row onMergedRows(Row merged, Row[] versions)
                 {
                     // cache the row versions to be able to regenerate the original row iterator
                     for (int i = 0; i < versions.length; i++)
+                    {
                         builders[i].addRow(versions[i]);
+                        rowsCached++;
+
+                        if (rowsCached == cachedRowsFailThreshold + 1)
+                        {
+                            throw new TooManyCachedRowsException(cachedRowsFailThreshold, command);
+                        }
+                        else if (rowsCached == cachedRowsWarnThreshold + 1)
+                        {
+                            String message =
+                                String.format("Replica filtering protection has cached over %d rows during query %s. " +
+                                              "(See 'cached_replica_rows_warn_threshold' in cassandra.yaml.)",
+                                              cachedRowsWarnThreshold, command.toCQLString());
+
+                            ClientWarn.instance.warn(message);
+                            logger.warn(message);
```

Comment:
@adelapena @jrwest Would this be a good candidate for the `NoSpamLogger`? I imagine we'd want to log all our failures, but between the thresholds, perhaps flooding the logs wouldn't be the best idea.